Cython/Compiler/Parsing.py

```diff
@@ -151,11 +151,25 @@ def p_test(s):
 
 #test_nocond: or_test | lambdef_nocond
 
-def p_test_nocond(s):
+def p_test_nocond(s, allow_assignment_expression=True):
     if s.sy == 'lambda':
         return p_lambdef_nocond(s)
     else:
-        return p_or_test(s)
+        return p_walrus_test(s, allow_assignment_expression)
+
+# walrurus_test: IDENT := test | or_test
+
+def p_walrus_test(s, allow_assignment_expression=True):
+    lhs = p_or_test(s)
+    if s.sy == ':=':
+        if not allow_assignment_expression:
+            s.error("invalid syntax: assignment expression not allowed in this context")
+        elif not lhs.is_name:
+            s.error("Left-hand side of assignment expression must be an identifier")
+        s.next()
+        rhs = p_test(s)
+        return ExprNodes.AssignmentExpressionNode(s.position(), lhs=lhs, rhs=rhs)
```

Comment:
This won't report the right position on later errors.
```suggestion
    if s.sy == ':=':
        position = s.position()
        if not allow_assignment_expression:
            s.error("invalid syntax: assignment expression not allowed in this context")
        elif not lhs.is_name:
            s.error("Left-hand side of assignment expression must be an identifier")
        s.next()
        rhs = p_test(s)
        return ExprNodes.AssignmentExpressionNode(position, lhs=lhs, rhs=rhs)
```